neqo-qpack/src/decoder.rs

```diff
@@ -526,206 +160,26 @@ impl QPackDecoder {
     // this function returns None if the stream is blocked waiting for table insertions.
     pub fn decode_header_block(&mut self, buf: &[u8], stream_id: u64) -> Res<Option<Vec<Header>>> {
         qdebug!([self], "decode header block.");
-        let mut reader = BufWrapper { buf, offset: 0 };
-
-        let (req_inserts, base) = self.read_base(&mut reader)?;
-        qdebug!(
-            [self],
-            "requested inserts count is {} and base is {}",
-            req_inserts,
-            base
-        );
-        if self.table.base() < req_inserts {
-            qdebug!(
-                [self],
-                "stream is blocked stream_id={} requested inserts count={}",
-                stream_id,
-                req_inserts
-            );
-            self.blocked_streams.push((stream_id, req_inserts));
-            if self.blocked_streams.len() > self.max_blocked_streams as usize {
-                return Err(Error::DecompressionFailed);
-            }
-            return Ok(None);
-        }
-        let mut h: Vec<Header> = Vec::new();
-
-        loop {
-            if reader.done() {
-                // Send header_ack
-                if req_inserts != 0 {
-                    self.header_ack(stream_id, req_inserts);
+        let mut decoder = HeaderDecoder::new(buf);
+
+        match decoder.decode_header_block(
+            &self.table,
+            self.max_entries,
+            self.total_num_of_inserts,
+        )? {
+            HeaderDecoderResult::Blocked(req_insert_cnt) => {
+                self.blocked_streams.push((stream_id, req_insert_cnt));
+                if self.blocked_streams.len() > self.max_blocked_streams as usize {
```

Comment:
I know that this code needs a bit more refactoring, I just needed to stop somewhere for now otherwise it would take me long to have this ready for review.
Thanks for all comments, I will fix them.

Reply:
That's fine.  Feel free to set some of these comments aside as future work.